frame/evm/src/resource.rs

```diff
@@ -0,0 +1,627 @@
+use crate::{AccountCodes, AccountCodesMetadata, Config, Pallet};
+
+use core::marker::PhantomData;
+use evm::{
+	gasometer::{GasCost, StorageTarget},
+	ExitError, Opcode,
+};
+use fp_evm::WeightInfo;
+use sp_core::{Get, H160, H256, U256};
+use sp_runtime::{traits::CheckedAdd, Saturating};
+use sp_std::{collections::btree_map::BTreeMap, vec::Vec};
+
+/// `System::Account` 16(hash) + 20 (key) + 60 (AccountInfo::max_encoded_len)
+pub const ACCOUNT_BASIC_PROOF_SIZE: u64 = 96;
+/// `AccountCodesMetadata` read, temptatively 16 (hash) + 20 (key) + 40 (CodeMetadata).
+pub const ACCOUNT_CODES_METADATA_PROOF_SIZE: u64 = 76;
+/// 16 (hash1) + 20 (key1) + 16 (hash2) + 32 (key2) + 32 (value)
+pub const ACCOUNT_STORAGE_PROOF_SIZE: u64 = 116;
+/// Fixed trie 32 byte hash.
+pub const WRITE_PROOF_SIZE: u64 = 32;
+/// Account basic proof size + 5 bytes max of `decode_len` call.
+pub const IS_EMPTY_CHECK_PROOF_SIZE: u64 = 93;
+/// Storage: Set a new value. (32 bytes)
+pub const STORAGE_NEW_COST: u64 = 32;
+
+#[derive(Debug, PartialEq)]
+/// Resource error.
+pub enum ResourceError {
+	/// The Resource usage exceeds the limit.
+	LimitExceeded,
+	/// Invalid Base Cost.
+	InvalidBaseCost,
+	///Used to indicate that the code should be unreachable.
+	Unreachable,
+}
+
+impl From<ResourceError> for ExitError {
+	fn from(_error: ResourceError) -> Self {
+		ExitError::OutOfGas
+	}
+}
+
+/// A struct that keeps track of resource usage and limit.
+pub struct Resource<T> {
+	limit: T,
+	usage: T,
+}
+
+impl<T> Resource<T>
+where
+	T: CheckedAdd + Saturating + PartialOrd + Copy,
+{
+	/// Creates a new `Resource` instance with the given base cost and limit.
+	///
+	/// # Errors
+	///
+	/// Returns `ResourceError::InvalidBaseCost` if the base cost is greater than the limit.
+	pub fn new(base_cost: T, limit: T) -> Result<Self, ResourceError> {
+		if base_cost > limit {
+			return Err(ResourceError::InvalidBaseCost);
+		}
+		Ok(Self {
+			limit,
+			usage: base_cost,
+		})
+	}
+
+	/// Records the cost of an operation and updates the usage.
+	///
+	/// # Errors
+	///
+	/// Returns `ResourceError::LimitExceeded` if the Resource usage exceeds the limit.
+	fn record_cost(&mut self, cost: T) -> Result<(), ResourceError> {
+		let usage = self
+			.usage
+			.checked_add(&cost)
+			.ok_or(ResourceError::LimitExceeded)?;
+
+		if usage > self.limit {
+			return Err(ResourceError::LimitExceeded);
+		}
```

Comment:
When this errors, it doesn't actually update `self.usage`.

I can see a good argument that it should work this way, but I also think it would be good defensive programming to update it (either leave it "invalidly" > `self.limit` or maybe just set `self.usage = self.limit`.)

In any case, it's probably good to document the behavior.

Reply:
Please check new version